internal/buff/read.go

```diff
@@ -0,0 +1,226 @@
+// This source file is part of the EdgeDB open source project.
+//
+// Copyright 2020-present EdgeDB Inc. and the EdgeDB authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package buff
+
+import (
+	"encoding/binary"
+	"fmt"
+	"reflect"
+	"unsafe"
+
+	"github.com/edgedb/edgedb-go/internal/soc"
+	"github.com/edgedb/edgedb-go/types"
+)
+
+// Reader is a buffer reader.
+type Reader struct {
+	toBeDeserialized chan *soc.Data
+
+	data    *soc.Data
+	Err     error
+	Buf     []byte
+	MsgType uint8
+}
+
+// NewReader returns a new Reader.
+func NewReader(toBeDeserialized chan *soc.Data) *Reader {
+	return &Reader{toBeDeserialized: toBeDeserialized}
+}
+
+// SimpleReader creates a new reader that operates on a single []byte.
+func SimpleReader(buf []byte) *Reader {
+	r := &Reader{Buf: buf}
+	(*reflect.SliceHeader)(unsafe.Pointer(&r.Buf)).Cap = len(buf)
+	return r
+}
+
+// Next advances the reader to the next message.
+// Next returns false when the reader doesn't own any socket data
+// and a signal is received on doneReadingSignal,
+// or an error is encountered while reading.
+//
+// Callers must continue to call Next until it returns false.
+//
+// If the previous message was not fully read Next() panics.
+//
+// Next() panics if called on a reader created with SimpleReader().
+func (r *Reader) Next(doneReadingSignal chan struct{}) bool {
+	if r.toBeDeserialized == nil {
+		panic("called next on a simple reader")
+	}
+
+	if len(r.Buf) > 0 {
+		panic(fmt.Sprintf(
+			"cannot finish: unread data in buffer (message type: 0x%x)",
+			r.MsgType,
+		))
+	}
+
+	if r.data != nil && len(r.data.Buf) == 0 {
+		r.data.Release()
+		r.data = nil
+	}
+
+	r.MsgType = 0
+
+	if r.data == nil {
+		select {
+		case <-doneReadingSignal:
+			return false
+		case r.data = <-r.toBeDeserialized:
+			if r.data.Err != nil {
+				r.Err = r.data.Err
+				r.data.Release()
+				r.data = nil
+				return false
+			}
+		}
+	}
+
+	r.Err = r.feed(5)
```

Comment:
What does `5` mean here?

Reply:
Put 5 bytes into `r.Buf` so that the message type and length can be read.

Reply:
Can you add a comment? Also how do you know that the buffer actually has 5 bytes in it?

Reply:
At this point we know that the server has sent more data so we know that there is another message to process. If there is a message to process then there must be at least 5 bytes that the server has sent or will send. `feed` waits on data from the `soc.Read` go routine if the reader has fewer than 5 bytes.

comment added :heavy_check_mark: